cobbler/actions/mkloaders.py

```diff
@@ -0,0 +1,212 @@
+"""Cobbler action to create bootable Grub2 images.
+
+This action calls grub2-mkimage for all bootloader formats configured in
+Cobbler's settings. See man(1) grub2-mkimage for available formats.
+"""
+import logging
+import pathlib
+import subprocess
+import sys
+import typing
+
+from cobbler import utils
+
+
+# NOTE: does not warrant being a class, but all Cobbler actions use a class's ".run()" as the entrypoint
+class MkLoaders:
+    """
+    Action to create bootloader images.
+    """
+
+    def __init__(self, api):
+        """
+        MkLoaders constructor.
+
+        :param api: CobblerAPI instance for accessing settings
+        """
+        self.logger = logging.getLogger()
+        self.bootloaders_dir = pathlib.Path(api.settings().bootloaders_dir)
+        # GRUB 2
+        self.grub2_mod_dir = pathlib.Path(api.settings().grub2_mod_dir)
+        self.boot_loaders_formats: typing.Dict = api.settings().bootloaders_formats
+        self.modules: typing.List = api.settings().bootloaders_modules
+        # Syslinux
+        self.syslinux_dir = pathlib.Path(api.settings().syslinux_dir)
+        self.syslinux_links = {
+            self.syslinux_dir.joinpath(f): self.bootloaders_dir.joinpath(f)
+            for f in ["pxelinux.0", "menu.c32", "ldlinux.c32", "memdisk"]
+        }
+
+    def run(self):
+        """
+        Run GrubImages action. If the files or executables for the bootloader is not available we bail out and skip the
+        creation after it is logged that this is not available.
+        """
+        self.create_directories()
+
+        self.make_shim()
+        self.make_ipxe()
+        self.make_syslinux()
+        self.make_grub()
+
+    def make_shim(self):
+        """
+        Create symlink of the shim bootloader in case it is available on the system.
+        """
+        if not utils.command_existing("shim-install"):
+            self.logger.info("shim-install missing. This means we are probably also missing the file we require. "
+                             "Bailing out of linking the shim!")
+            return
+        symlink(
+            pathlib.Path("/usr/share/efi/x86_64/shim.efi"),
+            self.bootloaders_dir.joinpath(pathlib.Path("undionly.pxe")),
+            skip_existing=True
+        )
+
+    def make_ipxe(self):
+        """
+        Create symlink of the iPXE bootloader in case it is available on the system.
+        """
+        if not pathlib.Path("/usr/share/ipxe").exists():
+            self.logger.info("ipxe directory did not exist. Bailing out of iPXE setup!")
+            return
+        symlink(
+            pathlib.Path("/usr/share/ipxe/undionly.kpxe"),
+            self.bootloaders_dir.joinpath(pathlib.Path("grub/shim.efi")),
```

Comment:
Aren't these `link` arguments the wrong way around?

Reply:
Done via 0fb04e0